### PR TITLE
ROX-17335: Reconcile auth provider + groups declaratively.

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This PR adds the capability to reconcile auth provider configuration declaratively instead of imperatively as done previously via separate API calls to Central.

In essence, the declarative config secret will now include an additional key of `default-sso-auth-provider` which contains all previously existing configuration in the form of declarative configuration.

Additionally, the previously used code was deleted which was now rendered unused after the refactoring.

One caveat exists within the PR:

For a limited time (namely until [ROX-17336](https://issues.redhat.com/browse/ROX-17336) has landed), we will have auth providers created in two different ways: the "legacy" way via API calls, as well as the "new" way via declarative config.

During that period, in case we reconcile Central with the "legacy" auth provider, we will create declarative configs for the auth providers, but the configuration will fail to be applied due to a name uniqueness violation.

This isn't as bad though, since we ensure that these errors will not be highlighted to the customer, and once we migrate we can do so with virtually no downtime for existing clients.

## Checklist (Definition of Done)
- [X] Unit and integration tests added
- [X] Added test description under `Test manual`
~~- [ ] Documentation added if necessary (i.e. changes to dev setup, test execution, ...)~~
- [X] CI and all relevant tests are passing
- [X] Add the ticket number to the PR title if available, i.e. `ROX-12345: ...`
~~- [ ] Discussed security and business related topics privately. Will move any security and business related topics that arise to private communication channel.~~
~~- [ ] Add secret to app-interface Vault or Secrets Manager if necessary~~
~~- [ ] RDS changes were e2e tested [manually](../docs/development/howto-e2e-test-rds.md)~~
~~- [ ] Check AWS limits are reasonable for changes provisioning new resources~~

## Test manual

```
# Start all relevant components (fleetmanager, fleet-sync, operator).
# Ensure that the fleetmanager specifies auth configuration (either the dynamic client SSO service account or the static IdP config) and that fleet-sync uses `CREATE_AUTHPROVIDER=true`.

1. Checkout the "main" branch and create a central instance via `./scripts/create-central.sh`

2. Observe that the central is created correctly and an auth provider is created (which is _not_ the basic auth provider) via
curl https://<central-endpoint>/v1/login/authproviders

3. Keep fleet manager running, stop fleet-sync, checkout the current branch, run `make binary` and start fleet-sync again.

4. Observe that reconciliation works without errors and the auth provider doesn't change. Additionally, observe that within the secret `sensible-declarative-configs` only a single data key exists, and `default-sso-auth-provider` does not.

5. Delete the previously created central with the "legacy" auth provider:
curl -X DELETE -H "Authorization: Bearer $(ocm token)" http://127.0.0.1:8000/api/rhacs/v1/centrals/<central-id>\?async\=true

6. Create a new Central via `./scripts/create-central.sh`.

7. Observe within fleet-sync logs that no error occurred. Additionally, observe that `sensible-declarative-configs` secret now contains the data key `default-sso-auth-provider`.

8. Ensure login works within the Central instance.

9. Delete the previously created central with the declarative auth provider:
curl -X DELETE -H "Authorization: Bearer $(ocm token)" http://127.0.0.1:8000/api/rhacs/v1/centrals/<central-id>\?async\=true

10. Ensure that the `sensible-declarative-configs` secret is removed and no config is retained of the auth provider.
```

### CODE REVIEW

File: fleetshard/pkg/central/reconciler/reconciler.go [hunk 1]
```diff
@@ -337,31 +352,172 @@ func (r *CentralReconciler) getInstanceConfig(remoteCentral *private.ManagedCent
...
 
+func (r *CentralReconciler) ensureAuthProviderExists(ctx context.Context, remoteCentral *private.ManagedCentral) error {
+	// Short-circuit if an auth provider isn't desired.
+	if !r.wantsAuthProvider {
+		return nil
+	}
+
+	// Since the auth provider is an integral part, we have to verify that it will be created successfully and listed
+	// within the list of available auth providers.
+	// This has to be done asynchronously now, since the declarative config will take time to be applied.
```
Comment (ivan-degtiarenko) [inline fleetshard/pkg/central/reconciler/reconciler.go:473]: I'm not sure asynchronously is the right word here, as the `concurrency.WaitInContext` blocks execution. Perhaps I would rephrase to something like:
```
We try multiple times to verify auth provider since applying of declarative configuration is done asynchronously.
```

File: fleetshard/pkg/central/reconciler/reconciler.go [hunk 2]
```diff
@@ -337,31 +352,172 @@ func (r *CentralReconciler) getInstanceConfig(remoteCentral *private.ManagedCent
...
+	authProviderExists := concurrency.WaitInContext(concurrency.NewPoller(func() bool {
+		exists, _, _ := r.verifyAuthProviderFunc(ctx, remoteCentral, r.client)
+		return exists
+	}, 5*time.Second), ctx)
+	if authProviderExists {
+		r.hasAuthProvider = true
+		return nil
+	}
+	return errors.New("failed to verify that the auth provider exists within Central API")
```
Comment (ivan-degtiarenko) [inline fleetshard/pkg/central/reconciler/reconciler.go:482]:
```suggestion
	return errors.New("failed to verify via Central API that the auth provider exists")
```

File: e2e/e2e_test.go [hunk 3]
```diff
@@ -192,6 +194,22 @@ var _ = Describe("Central", func() {
...
 		})
 
+		It("should create the secret containing declarative configurations", func() {
+			if createdCentral == nil {
+				Fail("central not created")
+			}
+
+			Eventually(func() error {
+				secret := &corev1.Secret{}
+				err := k8sClient.Get(context.TODO(), ctrlClient.ObjectKey{Name: declarativeConfigSecretName,
```
Comment (ivan-degtiarenko) [inline e2e/e2e_test.go:204]: Should we verify that there's auth provider data in this secret?

File: fleetshard/pkg/central/reconciler/util.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	address, err := getServiceAddress(ctx, central, client)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			return false, false, nil
+		}
+		return false, false, err
+	}
+
+	centralClient := centralClientPkg.NewCentralClientNoAuth(*central, address)
```
[the remark this answers is not in the document]
Comment (dhaus67) [reply fleetshard/pkg/central/reconciler/util.go:101]: I think it's worthwhile changing the signature, but not for `NewCentralClient..` but rather `isCentralDeploymentReady`, `getServiceAddress`, and `authProviderName`. They don't require a pointer being passed.

File: fleetshard/pkg/central/reconciler/reconciler.go [hunk 5]
```diff
@@ -337,31 +352,172 @@ func (r *CentralReconciler) getInstanceConfig(remoteCentral *private.ManagedCent
...
+	if err != nil {
+		return fmt.Errorf("failed to verify whether the sso.r.c auth provider exists: %w", err)
 	}
 
-	if r.hasAuthProvider {
-		central.Spec.Central.AdminPasswordGenerationDisabled = pointer.Bool(true)
+	// Short-circuit in case of a legacy auth provider and do not attempt to create declarative config ones since the
+	// creation would fail.
+	// TODO(ROX-17336): Once the legacy auth providers are migrated, remove this explicit check.
+	if legacyAuthProvider {
```
[the remark this answers is not in the document]
Comment (dhaus67) [reply fleetshard/pkg/central/reconciler/reconciler.go:372]: `isLegacyAuthProvider` can only be true if `exists == true`, but yeah I can add the exists to the condition as well to make it more readable.

File: fleetshard/pkg/central/reconciler/reconciler.go [hunk 6]
```diff
@@ -295,6 +305,11 @@ func (r *CentralReconciler) getInstanceConfig(remoteCentral *private.ManagedCent
 							Name: sensibleDeclarativeConfigSecretName,
 						},
 					},
+					ConfigMaps: []v1alpha1.LocalConfigMapReference{
```
[the remark this answers is not in the document]
Comment (dhaus67) [reply fleetshard/pkg/central/reconciler/reconciler.go:308]: I'd assume this needs to be done for the parent PR as well.

When testing this locally, the operator rendered the CRD as invalid with the message that `declarativeConfiguration.configMaps.name` is required.

I think the underlying reason is that [within the config struct](https://github.com/stackrox/stackrox/blob/master/operator/apis/platform/v1alpha1/central_types.go#L176-L184) we are not using a pointer reference, and thus an empty array is rendered which violates the requirements 🤔 

I'm currently checking with Draco team to see whether it's possible to retroactively fix this, however for the time being it looks to be required (earliest we could remove this would be a new release of the operator, i.e. 4.2).

File: fleetshard/pkg/central/reconciler/reconciler.go [hunk 7]
```diff
@@ -337,31 +352,172 @@ func (r *CentralReconciler) getInstanceConfig(remoteCentral *private.ManagedCent
...
+			if err != nil {
+				return fmt.Errorf("checking whether auth provider config requires reconciliation: %w", err)
+			}
+
+			if !requiresReconciliation {
+				glog.Info("Auth provider configuration is already up to date.")
+				return nil
+			}
+
+			glog.Info("Found changes for auth provider configuration, applying them.")
```
[the remark this answers is not in the document]
Comment (dhaus67) [reply fleetshard/pkg/central/reconciler/reconciler.go:449]: Could be inferred, but I think it's better to directly add those logs since we already have the info in most of the cases.

File: fleetshard/pkg/central/reconciler/reconciler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dhaus67) [reply fleetshard/pkg/central/reconciler/reconciler.go]: Scratch that.

I tested locally with `make -C operator run` which throws an error when attempting to create the CR:
```
Central.platform.stackrox.io "test-central-1" is invalid: spec.central.declarativeConfiguration.configMaps: Required value
```

However, when using an operator installed via catalog source, the CR creation works..

I'll clarify with Draco why this doesn't work for `make -C operator run`, but reverted the change for now since it's not required.

File: fleetshard/pkg/central/reconciler/reconciler.go [hunk 8]
```diff
@@ -685,6 +814,22 @@ func (r *CentralReconciler) collectReconciliationStatus(ctx context.Context, rem
...
+func (r *CentralReconciler) ensureCentralDeclarativeConfigSecretCleaned(ctx context.Context, namespace string) error {
+	declarativeConfigSecret := &corev1.Secret{}
+
+	if err := r.client.Get(ctx, ctrlClient.ObjectKey{Namespace: namespace, Name: sensibleDeclarativeConfigSecretName},
+		declarativeConfigSecret); err != nil {
+		if apiErrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to retrieve declarative config secret %s/%s: %w",
+			sensibleDeclarativeConfigSecretName, namespace, err)
```
Comment (ivan-degtiarenko) [inline fleetshard/pkg/central/reconciler/reconciler.go:826]:
```suggestion
			namespace, sensibleDeclarativeConfigSecretName, err)
```
Worth checking in other places

File: fleetshard/pkg/central/reconciler/reconciler.go [hunk 9]
```diff
@@ -685,6 +814,22 @@ func (r *CentralReconciler) collectReconciliationStatus(ctx context.Context, rem
...
+	}
+
+	return errors.Wrapf(r.client.Delete(ctx, declarativeConfigSecret),
+		"failed to delete declarative config secret %s/%s", sensibleDeclarativeConfigSecretName, namespace)
```
Comment (ivan-degtiarenko) [inline fleetshard/pkg/central/reconciler/reconciler.go:830]:
```suggestion
		"failed to delete declarative config secret %s/%s", namespace, sensibleDeclarativeConfigSecretName)
```

**Review by ivan-degtiarenko — APPROVED** [review]
LGTM modulo error format

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1156** <a href="https://app.graphite.dev/github/pr/stackrox/acs-fleet-manager/1156" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/acs-fleet-manager/1156?utm_source=stack-comment).
